### PR TITLE
Refactor workspace server plugin

### PR DESCRIPTION
**Pull request by ruanyl**
### Description

- move `core/server/workspaces` to `plugins/workspace/server`
- move `core/public/workspace/workspaces_client` to `plugins/workspace/public`
- rename `workspaces` -> `workspace`

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by SuZhou-Joe — APPROVED** [review]

**Review by raintygao — APPROVED** [review]

File: src/core/public/workspace/workspaces_service.ts [hunk 1]
```diff
@@ -2,50 +2,54 @@
...
-  private client?: WorkspacesClientContract;
-  private formatUrlWithWorkspaceId(url: string, id: string) {
-    return url;
-  }
-  private setFormatUrlWithWorkspaceId(formatFn: WorkspacesStart['formatUrlWithWorkspaceId']) {
-    this.formatUrlWithWorkspaceId = formatFn;
-  }
-  public async setup({ http }: { http: HttpSetup; uiSettings: IUiSettingsClient }) {
-    this.client = new WorkspacesClient(http);
+export interface WorkspaceAttribute {
```
[the remark this answers is not in the document]
Comment (ruanyl) [reply src/core/public/workspace/workspaces_service.ts:19]: It's exported from `core/public` and used by workspace plugin, is that what you mean?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 2]
```diff
@@ -187,7 +187,7 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
 
   componentDidMount() {
     this._isMounted = true;
-    this.props.workspaces.client.currentWorkspaceId$.subscribe((workspaceId) =>
+    this.props.workspaces.currentWorkspaceId$.subscribe((workspaceId) =>
```
[the remark this answers is not in the document]
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:190]: the unsubscribe included in #76

**Review by Hailong-am — APPROVED** [review]

**Review by wanglam — APPROVED** [review]

File: src/plugins/workspace/server/index.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import { WorkspacePlugin } from './plugin';
+
+// This exports static code and TypeScript types,
+// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
+
+export function plugin(initializerContext: PluginInitializerContext) {
+  return new WorkspacePlugin(initializerContext);
+}
+
+export { MlCommonsPluginSetup, MlCommonsPluginStart } from './types';
```
Comment (Hailong-am) [inline src/plugins/workspace/server/index.ts:18]: we should remove this,right?

Comment (ruanyl) [reply src/plugins/workspace/server/index.ts:18]: That's right, forgot to clean it up, I will do it in next PR.

**Pull request merged**